### PR TITLE
Use pytest to test the output.

### PR DESCRIPTION
**Pull request by gabi-250**
This replaces the old test runner with `pytest`.

There are two types of tests: 
* Tests which run a few test programs through my runtime and assert that the output is the same as the output produced by running the binary obtained by compiling them with `clang`. These are located under `tests`.
* Tests which run `MarkUnoptimizedPass` on each of the test files to make sure `__unopt_` functions return a special value, in order to be able to check that the runtime actually transfers control to the `__unopt_` functions. At the moment, there is only one test of this type, and it can be found in `tests/test_control_flow`

### CODE REVIEW

File: src/tests/test_control_flow/Makefile [hunk 1]
```diff
@@ -0,0 +1,41 @@
+CC := clang
+LLC := ../../llvm/build/bin/llc
+CPOINTPASS := -Xclang -load -Xclang ../../passes/build/function_passes/libCheckPointPass.so
```
Comment (vext01) [inline src/tests/test_control_flow/Makefile]: Why don't you use a variable for the path prefix here? It would make things easier to read.

File: src/tests/test_control_flow/Makefile [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+UNOPTPASS := -Xclang -load -Xclang ../../passes/build/function_passes/libUnoptimizedCopyPass.so
+LIVEVARPASS := -Xclang -load -Xclang ../../passes/build/function_passes/libLiveVariablesPass.so
+MARKPASS := -Xclang -load -Xclang ./passes/build/basic_block_passes/libMarkUnoptimizedPass.so
+PASSFLAGS := $(CPOINTPASS) $(LIVEVARPASS) $(UNOPTPASS)
+STMAP_CHECKER_DIR := ../../stackmap_checker/
+OBJ_NAMES := utils.o stmap.o jump.o guard.o call_stack_state.o
+OBJS := $(foreach obj, $(OBJ_NAMES), $(STMAP_CHECKER_DIR)$(obj))
+PREFIX := trace
```
Comment (vext01) [inline src/tests/test_control_flow/Makefile]: Although it doesn't matter in this scenario, `PREFIX` is a special variable used in `install` targets. You probably want to avoid that name.

File: src/tests/test_control_flow/Makefile [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+TARGET_OBJS := $(foreach bin, $(EXECUTABLES), $(bin).o)
+
+.PHONY: all clean stackmap_checker
+
+all: stackmap_checker $(EXECUTABLES)
+
+stackmap_checker:
+	cd $(STMAP_CHECKER_DIR) && $(MAKE)
+
+.SECONDEXPANSION:
```
Comment (vext01) [inline src/tests/test_control_flow/Makefile]: Never seen this before! Why is this needed?

File: src/tests/test_control_flow/test_output.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+TEST_FILES = [os.path.splitext(name)[0] for name in
+                extract_files_with_extension(cur_dir, '.c')]
+
+
+def setup_module(module):
+    subprocess.run('cd {:} && make clean && make'.format(cur_dir), shell=True,
+                   stdout=subprocess.PIPE)
+
+
+@pytest.mark.parametrize('name', TEST_FILES)
```
Comment (vext01) [inline src/tests/test_control_flow/test_output.py]: Nice! I like it.

File: src/tests/test_output.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import os
+import argparse
+import subprocess
+import pytest
+
+
+def extract_files_with_extension(dir_name, extension):
```
Comment (vext01) [inline src/tests/test_output.py]: This is duplicated code. It could live in a `support.py` (or similar name) file and be imported?

Comment (vext01) [reply src/tests/test_output.py]: Same for any other duplication.

File: src/tests/test_output.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+def setup_module(module):
+    subprocess.run('cd {:} && make clean && make'.format(cur_dir), shell=True,
+                   stdout=subprocess.PIPE)
+
+
+@pytest.mark.parametrize('name', TEST_FILES)
+def test_output(name):
+    bin_path = '{test_dir}/{name}'.format(test_dir=cur_dir, name=name)
+    p = subprocess.run(bin_path, shell=True, stdout=subprocess.PIPE,
```
Comment (vext01) [inline src/tests/test_output.py]: You are not using `stderr`. Did you want to print it in your error message? If not you can omit `stderr=`

File: src/tests/test_control_flow/Makefile (hunk 3 above)
Comment (gabi-250) [reply src/tests/test_control_flow/Makefile]: It's because I use the `$@.o` variable in the prerequisite of the `$(EXECUTABLES)` rule. Secondary expansion is needed because automatic variables are not normally available in prerequisites. I am escaping `$@.o` (`$$@.o`) and enabling secondary expansion to tell `make` to perform a second round of expansions on the prerequisites of every rule, so that I can actually access `$@`, as described here: http://make.mad-scientist.net/secondary-expansion/

File: src/tests/test_output.py (hunk 6 above)
Comment (gabi-250) [reply src/tests/test_output.py]: Good point!

File: src/tests/test_control_flow/Makefile (hunk 2 above)
Comment (gabi-250) [reply src/tests/test_control_flow/Makefile]: Ah, I never knew that!

**Comment by gabi-250** [comment]
@vext01 Thank you for your feedback! 

I think I've addressed your comments, so this should be ready for a re-review.

**Comment by gabi-250** [comment]
@vext01 No, I just moved the test files (the C programs) to a separate folder.

**Comment by gabi-250** [comment]
And I renamed one of the test runners, because I wanted them to have unique names.

**Comment by vext01** [comment]
OK, if you sync the branch and  squash I'll merge this at first opportunity.

**Comment by gabi-250** [comment]
@vext01 All done.
